### DISCUSSION TITLE
Any flags for failsafe or lost frame in CRSF protocol ?

### DISCUSSION THREAD

**Discussion by hw999** [ZZ-Cat/CRSFforArduino#51] [General]
I would like to translate CRSF-channel information into a SBUS-data stream. The SBUS protocol has flags for lost frames and failsafe. I wonder : are there any such flags in the CRSF-protocol and if yes - are they implemented in the library? At first glance I did not discover anything that would show how FS and LF is handled.
Could you please spend a few sentences on how CRSF handles LF and FS ?

**Comment by ZZ-Cat** [comment]
The CRSF Protocol doesn't use flags for detecting failsafe nor has a counter for invalid frames. It uses link statistics in the telemetry stream from the receiver to the "flight controller", after the RC Channels are sent.

I only _just_ realised that I haven't coded in parsing of the Link Stats packet yet. So, I added it to the list in #49.

So, once I implement that, you would be to write up in your firmware some logic that watches the LQ and RSNR values, set thresholds for them, and detect failsafe that way.

As for invalid packets, there already is a CRC performed on each packet.  
CRSF for Arduino is coded to only pass _valid_ packets to your sketch.  
All of this happens internally when you call `update()` in your sketch.


If I _did_ implement a means of counting invalid packets, it would most likely be used for debugging purposes.

**Reply by ZZ-Cat** [reply]
I also just remembered that I have another project that uses the CRSF Protocol, and I wrote logic in for handling fail-safe (by watching the Link Statistics telemetry).

So, I can go ahead and port the logic over from that into CRSF for Arduino. =^/.~=

**Comment by ZZ-Cat** [comment]
I put this into #69. So, it's now in my Version 1.0.0 development pipeline (which is being tracked by #61).
